### PR TITLE
Fix testDanglingIndicesCanBeListed

### LINKED ISSUE #108288: [CI] DanglingIndicesRestIT.testDanglingIndicesCanBeDeleted failing
### CI Link

https://buildkite.com/elastic/elasticsearch-periodic-platform-support/builds/2637#018f4c10-dc4e-48be-8ee7-ba0ea1a047c1

### Repro line

./gradlew ':qa:smoke-test-http:javaRestTest' --tests "org.elasticsearch.http.DanglingIndicesRestIT.testDanglingIndicesCanBeDeleted" -Dtests.seed=BAF5712830721EF6 -Dtests.locale=fi-FI -Dtests.timezone=NZ -Druntime.java=21

### Does it reproduce?

No

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

org.elasticsearch.http.DanglingIndicesRestIT > testDanglingIndicesCanBeDeleted FAILED
--
  | java.lang.AssertionError:
  | Expected: <3>
  | but: was <2>
  | at __randomizedtesting.SeedInfo.seed([BAF5712830721EF6:164B422ACB257E03]:0)
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  | at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
  | at org.elasticsearch.test.ESTestCase.assertThat(ESTestCase.java:2160)
  | at org.elasticsearch.http.DanglingIndicesRestIT.listDanglingIndexIds(DanglingIndicesRestIT.java:176)
  | at org.elasticsearch.http.DanglingIndicesRestIT.testDanglingIndicesCanBeDeleted(DanglingIndicesRestIT.java:166)

### PR DESCRIPTION
**Pull request by idegtiarenko**
I believe the test started failing because of the recent changes to allow closing (and deleting shards) asyncronously. As a result dandling index API now is seeing a directory in partially deleted state, fails to interpret partial data and fails as a result.

There are 3 possible ways to fix it:
* ignore exceptions that are thrown when attempting to read index partial metadata (this is tricky as we could catch various types of exceptions)
* attempt to do the read under the all shards lock (this is tricky as we do not have index name when attempting to read the index metadata from disk)
* assert busy the API call to retry on the client level

Closes: https://github.com/elastic/elasticsearch/issues/108288

Related to: https://github.com/elastic/elasticsearch/pull/108145, https://github.com/elastic/elasticsearch/pull/108096

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Could we log the exceptions we're dropping here, just in case it hides something more interesting?

File: server/src/main/java/org/elasticsearch/gateway/MetaStateService.java [hunk 1]
```diff
@@ -172,20 +173,32 @@ List<IndexMetadata> loadIndicesStates(Predicate<String> excludeIndexPathIdsPredi
...
-                logger.debug("[{}] failed to find metadata for existing index location", indexFolderName);
+            } catch (IOException e) {
+                // index directory was deleted concurrently, ignoring
+            } catch (ElasticsearchException e) {
+                if ((e.getCause() instanceof IOException) == false) {
+                    throw e;
+                }
+            } catch (IllegalStateException e) {
+                // partially deleted index, ignoring
             }
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/gateway/MetaStateService.java:202]: I am not very happy on how this looks.

I wonder if we should instead try to find locks with the same index UUID and acquire if they are present?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/MetaStateService.java:202]: I don't think that'll solve the problem, the lock might not exist when you check, but then come back into existence to delete the data shortly after

**Review by DaveCTurner — APPROVED** [review]
LGTM, one suggestion

**Pull request merged**